source/extensions/filters/common/rbac/engine_impl.cc

```diff
@@ -80,6 +102,68 @@ bool RoleBasedAccessControlEngineImpl::checkPolicyMatch(
   return matched;
 }
 
+RoleBasedAccessControlMatcherEngineImpl::RoleBasedAccessControlMatcherEngineImpl(
+    const xds::type::matcher::v3::Matcher& matcher,
+    Server::Configuration::ServerFactoryContext& factory_context,
+    ActionValidationVisitor& validation_visitor, const EnforcementMode mode)
+    : mode_(mode) {
+  ActionContext context{false};
+  Envoy::Matcher::MatchTreeFactory<Matching::MatchingData, ActionContext> factory(
+      context, factory_context, validation_visitor);
+  matcher_ = factory.create(matcher)();
+  has_log_ = context.has_log_;
+
+  if (!validation_visitor.errors().empty()) {
+    // TODO(snowp): Output all violations.
+    throw EnvoyException(fmt::format("requirement violation while creating route match tree: {}",
+                                     validation_visitor.errors()[0]));
+  }
+}
+
+bool RoleBasedAccessControlMatcherEngineImpl::handleAction(const Network::Connection& connection,
+                                                           StreamInfo::StreamInfo& info,
+                                                           std::string* effective_policy_id) const {
+  return handleAction(connection, *Http::StaticEmptyHeaders::get().request_headers, info,
+                      effective_policy_id);
+}
+
+bool RoleBasedAccessControlMatcherEngineImpl::handleAction(
+    const Network::Connection& connection, const Envoy::Http::RequestHeaderMap& headers,
+    StreamInfo::StreamInfo& info, std::string* effective_policy_id) const {
+  Matching::MatchingDataImpl data(connection, headers, info);
+  const auto& result = Envoy::Matcher::evaluateMatch<Matching::MatchingData>(*matcher_, data);
+  ASSERT(result.match_state_ == Envoy::Matcher::MatchState::MatchComplete);
+  if (result.result_) {
+    auto action = result.result_()->getTyped<Action>();
+    if (effective_policy_id != nullptr) {
+      *effective_policy_id = action.name();
+    }
+
+    // If there is at least an LOG action in matchers, we have to turn on and off for shared log
+    // metadata every time when there is a connection or request.
+    auto rbac_action = action.action();
+    if (has_log_ && mode_ != EnforcementMode::Shadow) {
+      ProtobufWkt::Struct log_metadata;
```

Comment:
This is the same for both old/new API, can we refactor it to a single function and reuse it?

Reply:
Done.